File: prime_power_recurrence/scripts/search_prime_powers.py
#!/usr/bin/env python3
"""Exact searches for F_m^r + F_ell^r = T_n^p.

The script is intentionally independent of Sage/Magma so that large bounded
searches can be run quickly with Python's arbitrary-precision integers. Magma
scripts in ../magma handle the arithmetic-geometry side.
"""

from __future__ import annotations

import argparse
import csv
from collections import defaultdict
from math import gcd
from pathlib import Path
from time import perf_counter


DEFAULT_PRIMES = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]


def parse_int_list(text: str) -> list[int]:
    return [int(x.strip()) for x in text.split(",") if x.strip()]


def fibonacci_list(nmax: int) -> list[int]:
    if nmax < 0:
        return []
    if nmax == 0:
        return [0]
    seq = [0, 1]
    for _ in range(2, nmax + 1):
        seq.append(seq[-1] + seq[-2])
    return seq


def tribonacci_list(nmax: int) -> list[int]:
    if nmax < 0:
        return []
    if nmax == 0:
        return [0]
    if nmax == 1:
        return [0, 1]
    seq = [0, 1, 1]
    for _ in range(3, nmax + 1):
        seq.append(seq[-1] + seq[-2] + seq[-3])
    return seq


def triple_gcd(a: int, b: int, c: int) -> int:
    return gcd(gcd(abs(a), abs(b)), abs(c))


def build_right_map(fib: list[int], r: int, mmax: int) -> dict[int, list[tuple[int, int]]]:
    fpow = [x**r for x in fib[: mmax + 1]]
    out: dict[int, list[tuple[int, int]]] = defaultdict(list)
    for m in range(mmax + 1):
        fm = fpow[m]
        for ell in range(m + 1):
            out[fm + fpow[ell]].append((ell, m))
    return out


def classify_solution(r: int, p: int, ell: int, m: int, n: int, fl: int, fm: int, tn: int, g: int) -> str:
    if tn == 0 and fm == 0 and fl == 0:
        return "zero"
    if fl == 0:
        if p == r and fm == tn:
            return "one-zero-side: p=r and F_m=T_n"
        return "one-zero-side"
    if tn in (1, 2, 4, 13, 81) or fm in (1, 2, 3, 8, 13) or fl in (1, 2, 3, 8, 13):
        if g > 1:
            return "small-value imprimitive"
        return "small-value"
    if g > 1:
        return "imprimitive"
    return "primitive-nonzero"


def search(r_values: list[int], p_values: list[int], mmax: int, nmax: int) -> list[dict[str, object]]:
    fib = fibonacci_list(mmax)
    trib = tribonacci_list(nmax)
    rows: list[dict[str, object]] = []

    right_maps = {r: build_right_map(fib, r, mmax) for r in r_values}

    for r in r_values:
        right = right_maps[r]
        for p in p_values:
            for n, tn in enumerate(trib):
                target = tn**p
                pairs = right.get(target)
                if pairs is None:
                    continue
                for ell, m in pairs:
                    fl, fm = fib[ell], fib[m]
                    g = triple_gcd(fl, fm, tn)
                    nonzero = fl != 0 and fm != 0 and tn != 0
                    primitive = g == 1
                    rows.append(
                        {
                            "r": r,
                            "p": p,
                            "ell": ell,
                            "m": m,
                            "n": n,
                            "F_ell": fl,
                            "F_m": fm,
                            "T_n": tn,
                            "value": target,
                            "gcd": g,
                            "primitive": int(primitive),
                            "nonzero": int(nonzero),
                            "classification": classify_solution(r, p, ell, m, n, fl, fm, tn, g),
                        }
                    )

    rows.sort(key=lambda row: (row["r"], row["p"], row["n"], row["m"], row["ell"]))
    return rows


def write_csv(path: Path, rows: list[dict[str, object]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fields = [
        "r",
        "p",
        "ell",
        "m",
        "n",
        "F_ell",
        "F_m",
        "T_n",
        "value",
        "gcd",
        "primitive",
        "nonzero",
        "classification",
    ]
    with path.open("w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=fields)
        writer.writeheader()
        writer.writerows(rows)


def print_summary(rows: list[dict[str, object]], max_rows: int) -> None:
    print(f"solutions={len(rows)}")
    primitive_nonzero = [row for row in rows if row["primitive"] == 1 and row["nonzero"] == 1]
    print(f"primitive_nonzero={len(primitive_nonzero)}")

    by_rp: dict[tuple[int, int], int] = defaultdict(int)
    by_class: dict[str, int] = defaultdict(int)
    for row in rows:
        by_rp[(int(row["r"]), int(row["p"]))] += 1
        by_class[str(row["classification"])] += 1

    print("class_counts:")
    for label in sorted(by_class):
        print(f"  {label}: {by_class[label]}")

    print("nonempty_(r,p)_counts:")
    for (r, p), count in sorted(by_rp.items()):
        print(f"  r={r}, p={p}: {count}")

    print("first_rows:")
    for row in rows[:max_rows]:
        print(
            "  "
            f"r={row['r']} p={row['p']} n={row['n']} m={row['m']} ell={row['ell']} "
            f"T_n={row['T_n']} F_m={row['F_m']} F_ell={row['F_ell']} "
            f"gcd={row['gcd']} {row['classification']}"
        )


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--r-list", default="2,3,5,7,11,13", help="comma-separated exponents r")
    parser.add_argument("--p-list", default=",".join(map(str, DEFAULT_PRIMES)), help="comma-separated prime exponents p")
    parser.add_argument("--m-max", type=int, default=300)
    parser.add_argument("--n-max", type=int, default=300)
    parser.add_argument("--csv", default="data/prime_power_search.csv")
    parser.add_argument("--max-rows", type=int, default=120)
    args = parser.parse_args()

    t0 = perf_counter()
    rows = search(parse_int_list(args.r_list), parse_int_list(args.p_list), args.m_max, args.n_max)
    elapsed = perf_counter() - t0
    write_csv(Path(args.csv), rows)
    print_summary(rows, args.max_rows)
    print(f"csv={args.csv}")
    print(f"elapsed_seconds={elapsed:.3f}")


if __name__ == "__main__":
    main()

File: prime_power_recurrence/magma/search_prime_powers.m
// Exact Magma search for F_m^r + F_ell^r = T_n^p.
//
// Adjust the constants below and run:
//   /Applications/Magma/magma magma/search_prime_powers.m

SetColumns(0);

RValues := [ 2, 3, 5, 7, 11, 13 ];
PValues := [ 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31 ];
MMax := 300;
NMax := 300;
OutputFile := "data/prime_power_magma.csv";

function FibonacciList(N)
    F := [ Integers() | 0 : i in [0..N] ];
    if N ge 1 then
        F[2] := 1;
    end if;
    for k in [2..N] do
        F[k + 1] := F[k] + F[k - 1];
    end for;
    return F;
end function;

function TribonacciList(N)
    T := [ Integers() | 0 : i in [0..N] ];
    if N ge 1 then
        T[2] := 1;
    end if;
    if N ge 2 then
        T[3] := 1;
    end if;
    for k in [3..N] do
        T[k + 1] := T[k] + T[k - 1] + T[k - 2];
    end for;
    return T;
end function;

function TripleGCD(a, b, c)
    return GCD(GCD(Abs(a), Abs(b)), Abs(c));
end function;

function ClassLabel(r, p, ell, m, n, fl, fm, tn, g)
    if tn eq 0 and fm eq 0 and fl eq 0 then
        return "zero";
    end if;
    if fl eq 0 then
        if p eq r and fm eq tn then
            return "one-zero-side: p=r and F_m=T_n";
        end if;
        return "one-zero-side";
    end if;
    if tn in {1, 2, 4, 13, 81} or fm in {1, 2, 3, 8, 13} or fl in {1, 2, 3, 8, 13} then
        if g gt 1 then
            return "small-value imprimitive";
        end if;
        return "small-value";
    end if;
    if g gt 1 then
        return "imprimitive";
    end if;
    return "primitive-nonzero";
end function;

function RightMap(F, r, M)
    A := AssociativeArray(Integers());
    FPow := [ F[k + 1]^r : k in [0..M] ];
    for m in [0..M] do
        for ell in [0..m] do
            v := FPow[m + 1] + FPow[ell + 1];
            if IsDefined(A, v) then
                pairs := A[v];
            else
                pairs := [* *];
            end if;
            Append(~pairs, <ell, m>);
            A[v] := pairs;
        end for;
    end for;
    return A;
end function;

F := FibonacciList(MMax);
T := TribonacciList(NMax);

out := Open(OutputFile, "w");
fprintf out, "r,p,ell,m,n,F_ell,F_m,T_n,value,gcd,primitive,nonzero,classification\n";

total := 0;
primitiveNonzero := 0;

for r in RValues do
    printf "Building right map for r=%o\n", r;
    right := RightMap(F, r, MMax);
    for p in PValues do
        countRP := 0;
        for n in [0..NMax] do
            target := T[n + 1]^p;
            if IsDefined(right, target) then
                for pair in right[target] do
                    ell := pair[1];
                    m := pair[2];
                    fl := F[ell + 1];
                    fm := F[m + 1];
                    tn := T[n + 1];
                    g := TripleGCD(fl, fm, tn);
                    primitive := g eq 1;
                    nonzero := fl ne 0 and fm ne 0 and tn ne 0;
                    label := ClassLabel(r, p, ell, m, n, fl, fm, tn, g);
                    fprintf out, "%o,%o,%o,%o,%o,%o,%o,%o,%o,%o,%o,%o,%o\n",
                        r, p, ell, m, n, fl, fm, tn, target, g,
                        primitive select 1 else 0,
                        nonzero select 1 else 0,
                        label;
                    total +:= 1;
                    countRP +:= 1;
                    if primitive and nonzero then
                        primitiveNonzero +:= 1;
                    end if;
                end for;
            end if;
        end for;
        if countRP gt 0 then
            printf "r=%o p=%o solutions=%o\n", r, p, countRP;
        end if;
    end for;
end for;

delete out;

printf "Total solutions: %o\n", total;
printf "Primitive nonzero solutions: %o\n", primitiveNonzero;
printf "Wrote %o\n", OutputFile;
